File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
from collections import deque

def solution(n, edge):
    
    # 1번 노드부터 각 노드까지의 거리를 저장 
    distances = [-1 for _ in range(n+1)]
    distances[1] = 0
    
    # edge 그래프(인접리스트) 만들기
    graph = [[] for _ in range(n+1)]
    
    for a,b in edge:
        graph[a].append(b)
        graph[b].append(a)
        
    # BFS에서 방문 예정인 노드 담기 
    q = deque([1])
    
    while q:
        cur = q.popleft()
        for neighbor in graph[cur]:
            if distances[neighbor] == -1:
                q.append(neighbor)
                distances[neighbor] = distances[cur] + 1
    
    
    
    most_far = max(distances)
    
    return distances.count(most_far)
        
        
    

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.cpp
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<vector<int>> edge) {
    // 0번 노드로부터 거리 저장
    vector<int> visited(n, -1);
    visited[0] = 0; //자기자신이므로 
    
    // 인접리스트 그래프 만들기
    vector<vector<int>> graph(n, vector<int>(0));
    
    for(vector<int> e: edge){
        //0번부터 시작하는 걸로 초기화
        int first = e[0]-1;
        int second = e[1]-1;
        
        graph[first].push_back(second);
        graph[second].push_back(first);
    }
    
    // 최단 경로 BFS
    queue<int> q;
    q.push(0); //0번 노드에서 시작
    
    while(!q.empty()){
        int cur = q.front();
        q.pop();
        
        for(int x: graph[cur]){
            if(visited[x] == -1) {
                visited[x] = visited[cur] + 1;
                q.push(x);
            }
        }
    }
    
    for(int g: visited){
        cout<< g <<" ";
    }
    cout << endl;
    
    // 가장 멀리 떨어진 거리
    int far_amt = *max_element(visited.begin(), visited.end());
    
    // far_amt만큼 떨어진 노드 반환
    return count(visited.begin(), visited.end(), far_amt);
}